README.md
### Forked from https://github.com/sourceclear/cron-builder

# cron-builder
The software utility *cron* is a time-based scheduler in Unix-like computer operating systems. A user may use cron to schedule jobs (commands or scripts) to run periodically at fixed times, dates, or intervals. A cron statement is composed of 5 fields separated by white space. The `*` character translates to "every", ie: "every minute" or "every day of the week".

cron-builder will manage the state of a cron expression, allowing a user to manipulate it through a simple API. It is decoupled from the DOM and doesn't have an opinion about where it's being called from. cron-builder considers [this article](https://en.wikipedia.org/wiki/Cron) its source of truth.

### Install
cron-builder is available on npm and bower:
```
npm i cron-builder-ts --save
```

After installing, just require the package as you normally would:
```
import { CronBuilder } from 'cron-builder-ts';
```



### API
To instantiate the cron builder:

```JavaScript
// (default expression is set to "* * * * *")
var cronExp = new CronBuilder();

// optionally, pass in a cron expression to override the default:
var myCronExp = new CronBuilder('5 12 * * 1-5')
```

To return the cron expression at any given time:
```JavaScript
cronExp.build();
// '* * * * *'
```

API includes basic getters and setters:
```JavaScript
cronExp.get('minute');
// '*'
cronExp.set('minute', '5,35');
// '5,35'
cronExp.get('minute');
// '5,35'
cronExp.build();
// '5,35 * * * *'
```

Or if you'd prefer to add or remove values one at a time, use `addValue` and `removeValue`. These methods build or take away from what is currently set:
```JavaScript
cronExp.addValue('hour', '2');
cronExp.addValue('monthOfTheYear', '4');
cronExp.addValue('monthOfTheYear', '10');
cronExp.build();
// '5,35 2 * 4,10 *'

cronExp.removeValue('minute', '5');
cronExp.build();
// '35 2 * 4,10 *'
```

If you prefer to work with the expression object directly, use `getAll` and `setAll`:
```JavaScript
var exp = cronExp.getAll();
// {minute: ['35'], hour: ['2'], dayOfTheMonth: ['*'], monthOfTheYear: ['4','10'], ...}
exp.dayOfTheMonth = ['7','14','21','28'];
cronExp.setAll(exp);
cronExp.build();
// '35 2 7,14,21,28 4,10 *'
```

##### Notes:
- cron-builder does not currently support using `/` syntax to indicate values that are repeated. Instead of using `*/15`, use the verbose form `0,15,30,45`.
- cron-builder requires using numeric representations of days of the week and months of the year. So instead of using `Feb,Mar,Apr` just use `2,3,4`.

### Contribute
Easy!
```
npm install
npm test
```
Pull requests and issues appreciated!

### TODO
- more validation for things like adding values that are already included in a range, better range validations...
- sorting of values for a given measure of time

### License

The MIT License (MIT)


CHANGELOG.md
<a name="1.0.5"></a>
# 1.0.3 (2020-09-04)


### Fix

* npm package

<a name="1.0.4"></a>
# 1.0.3 (2020-09-04)


### Fix

* cron-converter dependency

<a name="1.0.3"></a>
# 1.0.3 (2020-09-04)


### Features

* using ([cron-converter](https://www.npmjs.com/package/cron-converter)) internally
* extend build method with cron-converter options
* extend get and getAll methods with 'expand' flag (false by default). if true returns full array of numbers for each filed in expression


projects/cron-builder-ts/src/lib/cron-builder.spec.ts
import { CronBuilder } from './cron-builder';

describe('cron-builder', () => {
  it('defaults to "* * * * *" when initialized without arguments', () => {
    const cron = new CronBuilder();
    expect(cron.get('minute')).toEqual('*');
    expect(cron.get('hour')).toEqual('*');
    expect(cron.get('dayOfTheMonth')).toEqual('*');
    expect(cron.get('month')).toEqual('*');
    expect(cron.get('dayOfTheWeek')).toEqual('*');
  });

  it('should split multiple cron values when initialized with arguments', () => {
    const cron = new CronBuilder("0,15,30,45 * * * *");
    const { minute } = cron.getAll();
    expect(minute).toEqual(["0", "15", "30", "45"]);
  });

  it('should split multiple cron values when initialized with arguments (expanded)', () => {
    const cron = new CronBuilder("0,15,30,45 * * * *");
    const {
      dayOfTheMonth,
      dayOfTheWeek,
      hour,
      minute,
      month,
    } = cron.getAll({ expand: true });

    expect(minute).toEqual([0, 15, 30, 45]);
    expect(hour).toEqual([0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23]);
    expect(dayOfTheMonth).toEqual([1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31]);
    expect(month).toEqual([1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12]);
    expect(dayOfTheWeek).toEqual([0, 1, 2, 3, 4, 5, 6]);
  });

  it('returns a working cron expression when calling .build()', () => {
    const cron = new CronBuilder();
    expect(cron.build()).toEqual('* * * * *');
  });

  it('sets a single value', () => {
    const cron = new CronBuilder();
    expect(cron.set('hour', ['5'])).toEqual('5');
    expect(cron.build()).toEqual('* 5 * * *');
  });

  it('sets multiple values at once', () => {
    const cron = new CronBuilder();
    expect(cron.set('minute', ['0', '10', '20', '30', '40', '50'])).toEqual('0,10,20,30,40,50');
    expect(cron.build({ plain: false })).toEqual('*/10 * * * *');
    expect(cron.build()).toEqual('0,10,20,30,40,50 * * * *');
  });

  it('sets a range', () => {
    const cron = new CronBuilder();
    expect(cron.set('dayOfTheWeek', ['5-7'])).toEqual('5-7');
    expect(cron.build({ plain: false })).toEqual('* * * * 0,5-6');
    expect(cron.build()).toEqual('* * * * 5-7');
  });

  it('multiple sets build the cron string accurately', () => {
    const cron = new CronBuilder();
    cron.set('minute', ['10', '30', '50']);
    cron.set('hour', ['6', '18']);
    cron.set('dayOfTheMonth', ['1', '15']);
    cron.set('dayOfTheWeek', ['1-5']);
    expect(cron.build({ plain: false })).toEqual('10-50/20 6,18 1,15 * 1-5');
    expect(cron.build()).toEqual('10,30,50 6,18 1,15 * 1-5');
  });

  it('validates against setting a value that is not a number or range of numbers', () => {
    const cron = new CronBuilder();
    expect(() => { cron.set('hour', ['!']) }).toThrow();
  });

  describe('validates against setting values outside the valid range', () => {
    it('validates against values too low', () => {
      const cron = new CronBuilder();
      expect(() => { cron.set('dayOfTheWeek', ['-1']) }).toThrow();
    });

    it('validates against values too high', () => {
      const cron = new CronBuilder();
      expect(() => { cron.set('hour', ['100']) }).toThrow();
    });

    it('validates against setting a range that is out of bounds', () => {
      const cron = new CronBuilder();
      expect(() => { cron.set('minute', ['20-60']) }).toThrow();

      expect(() => { cron.set('hour', ['12', '22-26', '15']) }).toThrow();
    });
  });

  it('gets a single value', () => {
    const cron = new CronBuilder();
    cron.set('minute', ['30']);
    expect(cron.get('minute')).toEqual('30');
  });

  it('returns the entire expression object when getAll is called', () => {
    const cron = new CronBuilder();
    const getAllResponse = cron.getAll();
    expect(getAllResponse.minute[0]).toEqual('*');
    expect(getAllResponse.hour[0]).toEqual('*');
    expect(getAllResponse.dayOfTheMonth[0]).toEqual('*');
    expect(getAllResponse.month[0]).toEqual('*');
    expect(getAllResponse.dayOfTheWeek[0]).toEqual('*');
  });

  it('sets the entire object when setAll is called', () => {
    const cron = new CronBuilder();
    const getAllResponse = cron.getAll();
    getAllResponse.hour = ['13'];
    getAllResponse.month = ['1-6'];
    getAllResponse.dayOfTheWeek = ['1,3,5,7'];
    cron.setAll(getAllResponse);
    expect(cron.build({ plain: false })).toEqual('* 13 * 1-6 0-1,3,5');
    expect(cron.build()).toEqual('* 13 * 1-6 1,3,5,7');
  });

  it('validates setting with an incorrect value with setAll', () => {
    const cron = new CronBuilder();
    const getAllResponse = cron.getAll();
    getAllResponse.hour = ['28'];
    expect(() => { cron.setAll(getAllResponse) }).toThrow();
  });

  it('adds a value to a measureOfTime that is set to "*"', () => {
    const cron = new CronBuilder();
    cron.addValue('minute', '5');
    expect(cron.get('minute')).toEqual('5');
    expect(cron.build()).toEqual('5 * * * *');
  });

  it('adds a value to a measure of time that has been set to a number', () => {
    const cron = new CronBuilder();
    cron.addValue('hour', '5');
    cron.addValue('hour', '10');
    expect(cron.get('hour')).toEqual('5,10');
  });

  it('validates duplicate values', () => {
    const cron = new CronBuilder();
    cron.addValue('dayOfTheMonth', '5');
    cron.addValue('dayOfTheMonth', '15');
    cron.addValue('dayOfTheMonth', '5');
    expect(cron.get('dayOfTheMonth')).toEqual('5,15');
  });

  it('validates an invalid value when adding', () => {
    const cron = new CronBuilder();
    expect(() => { cron.addValue('minute', '62') }).toThrow();
  });

  it('removes a value that exists with other values', () => {
    const cron = new CronBuilder();
    cron.set('dayOfTheWeek', ['2', '4']);
    cron.removeValue('dayOfTheWeek', '4');
    expect(cron.get('dayOfTheWeek')).toEqual('2');
  });

  it('resets the value to the default "*" when removing the only value', () => {
    const cron = new CronBuilder();
    cron.set('minute', ['7']);
    expect(cron.get('minute')).toEqual('7');
    cron.removeValue('minute', '7');
    expect(cron.get('minute')).toEqual('*');
  });

  it('accepts a cron expression when instantiating', () => {
    const cron = new CronBuilder('30 0-6 * * 1-5');
    expect(cron.build()).toEqual('30 0-6 * * 1-5');
  });

  it('validates bad values when instantiating with an explicit expression', () => {
    expect(() => {
      const cron = new CronBuilder('30 0-6 * * 1-10');
      cron.build();
    }).toThrow();
  });

  it('validates an expression that is too long when instantiating with an explicit expression', () => {
    expect(() => {
      const cron = new CronBuilder('30 0-6 * * 1-5 * *');
      cron.build();
    }).toThrow();
  });
});


projects/cron-builder-ts/karma.conf.js
// Karma configuration file, see link for more information
// https://karma-runner.github.io/1.0/config/configuration-file.html

module.exports = function (config) {
  config.set({
    basePath: '',
    frameworks: ['jasmine', '@angular-devkit/build-angular'],
    plugins: [
      require('karma-jasmine'),
      require('karma-chrome-launcher'),
      require('karma-jasmine-html-reporter'),
      require('karma-coverage-istanbul-reporter'),
      require('@angular-devkit/build-angular/plugins/karma')
    ],
    client: {
      clearContext: false // leave Jasmine Spec Runner output visible in browser
    },
    coverageIstanbulReporter: {
      dir: require('path').join(__dirname, '../../coverage/cron-builder-ts'),
      reports: ['html', 'lcovonly', 'text-summary'],
      fixWebpackSourcePaths: true
    },
    reporters: ['progress', 'kjhtml'],
    port: 9876,
    colors: true,
    logLevel: config.LOG_INFO,
    autoWatch: true,
    browsers: ['Chrome'],
    singleRun: false,
    restartOnFileChange: true
  });
};


projects/cron-builder-ts/src/lib/types.ts
export const MINUTE = "minute";
export const HOUR = "hour";
export const DAY_OF_THE_MONTH = "dayOfTheMonth";
export const MONTH = "month";
export const DAY_OF_THE_WEEK = "dayOfTheWeek";

export type MeasureOfTime = typeof MINUTE | typeof HOUR | typeof DAY_OF_THE_MONTH | typeof MONTH | typeof DAY_OF_THE_WEEK;

export interface Expression {
  [MINUTE]: string[];
  [HOUR]: string[];
  [DAY_OF_THE_MONTH]: string[];
  [MONTH]: string[];
  [DAY_OF_THE_WEEK]: string[];
}

export interface ExpandedExpression {
  [MINUTE]: number[];
  [HOUR]: number[];
  [DAY_OF_THE_MONTH]: number[];
  [MONTH]: number[];
  [DAY_OF_THE_WEEK]: number[];
}

export const DEFAULT_INTERVAL = ['*'];

/**
 * Contains the position-to-name mapping of the cron expression
 * @type {Array}
 * @const
 */
export const MEASURE_OF_TIME_MAP: ReadonlyArray<MeasureOfTime> = [
  MINUTE,
  HOUR,
  DAY_OF_THE_MONTH,
  MONTH,
  DAY_OF_THE_WEEK,
];


projects/cron-builder-ts/src/lib/cron-validator.ts
import { Expression, MEASURE_OF_TIME_MAP } from './types';
import { keys } from './utils';

export class CronValidator {
  /**
   * validates a given cron expression (object) for length, then calls validateValue on each value
   * @param {!{
   *  minute: Array.string,
   *  hour: Array.string,
   *  dayOfTheMonth: Array.string,
   *  month: Array.string,
   *  dayOfTheWeek: Array.string,
   * }} expression - rich object containing the state of the cron expression
   * @throws {Error} if expression contains more than 5 keys
   */
  public static validateExpression(expression: Expression): void {
    // don't care if it's less than 5, we'll just set those to the default '*'
    if (keys(expression).length > MEASURE_OF_TIME_MAP.length) {
      throw new Error(`Invalid cron expression; limited to ${MEASURE_OF_TIME_MAP.length} values.`);
    }

    keys(expression).forEach((key) => {
      expression[key].forEach((value) => {
        CronValidator.validateValue(key, value);
      });
    });
  }

  /**
   * validates a given cron expression (string) for length, then calls validateValue on each value
   * @param {!String} expression - an optionally empty string containing at most 5 space delimited expressions.
   * @throws {Error} if the string contains more than 5 space delimited parts.
   */
  public static validateString(expression: string): void {
    const splitExpression = expression.split(' ');

    if (splitExpression.length > MEASURE_OF_TIME_MAP.length) {
      throw new Error(`Invalid cron expression; limited to ${MEASURE_OF_TIME_MAP.length} values.`);
    }

    for (let i = 0; i < splitExpression.length; i += 1) {
      CronValidator.validateValue(MEASURE_OF_TIME_MAP[i], splitExpression[i]);
    }
  }

  /**
   * validates any given measureOfTime and corresponding value
   * @param {!String} measureOfTime - as expected
   * @param {!String} value - the cron-ish interval specifier
   * @throws {Error} if measureOfTime is bogus
   * @throws {Error} if value contains an unsupported character
   */
  public static validateValue(measureOfTime: keyof Expression, value: string): void {
    const validatorObj = {
      minute: { min: 0, max: 59 },
      hour: { min: 0, max: 23 },
      dayOfTheMonth: { min: 1, max: 31 },
      month: { min: 1, max: 12 },
      dayOfTheWeek: { min: 0, max: 7 }
    };

    const validChars = /^[0-9*-]/;

    if (!validatorObj[measureOfTime]) {
      throw new Error(`Invalid measureOfTime; Valid options are: ${MEASURE_OF_TIME_MAP.join(', ')}`);
    }

    if (!validChars.test(value)) {
      throw new Error('Invalid value; Only numbers 0-9, "-", and "*" chars are allowed');
    }

    if (value === '*') {
      return;
    }

    // check to see if value is within range if value is not '*'
    if (value.indexOf('-') >= 0) {
      // value is a range and must be split into high and low
      const rangeArray: ReadonlyArray<number | undefined> = value
        .split('-')
        .map((range) => parseInt(range, 10))
        .filter((range) => isNaN(range) === false);

      const [rangeMin, rangeMax] = rangeArray;

      if (typeof rangeMin !== "number" || rangeMin < validatorObj[measureOfTime].min) {
        throw new Error(`Invalid value; bottom of range is not valid for \\"${measureOfTime}\\". Limit is ${validatorObj[measureOfTime].min}.`);
      }

      if (typeof rangeMax !== "number" || rangeMax > validatorObj[measureOfTime].max) {
        throw new Error(`Invalid value; top of range is not valid for \\"${measureOfTime}\\". Limit is ${validatorObj[measureOfTime].max}.`);
      }
    } else {
      if (parseInt(value) < validatorObj[measureOfTime].min) {
        throw new Error(`Invalid value; given value is not valid for \\"${measureOfTime}\\". Minimum value is \\"${validatorObj[measureOfTime].min}\\".`);
      }

      if (parseInt(value) > validatorObj[measureOfTime].max) {
        throw new Error(`Invalid value; given value is not valid for \\"${measureOfTime}\\". Maximum value is \\"${validatorObj[measureOfTime].max}\\".`);
      }
    }
  }
}


projects/cron-builder-ts/src/lib/cron-builder.ts
import * as Cron from "cron-converter";

import { CronValidator } from './cron-validator';
import { DEFAULT_INTERVAL, ExpandedExpression, Expression, MEASURE_OF_TIME_MAP } from './types';
import { keys } from './utils';

/**
 * Initializes a CronBuilder with an optional initial cron expression.
 * @param {String=} initialExpression - if provided, it must be up to 5 space delimited parts
 * @throws {Error} if the initialExpression is bogus
 * @constructor
 */
export class CronBuilder {
  private readonly expression: Expression;

  constructor(initialExpression: string = "* * * * *") {
    if (initialExpression) {
      CronValidator.validateString(initialExpression);

      const splitExpression = initialExpression.split(' ');

      this.expression = {
        minute: splitExpression[0] ? splitExpression[0].split(",") : DEFAULT_INTERVAL,
        hour: splitExpression[1] ? splitExpression[1].split(",") : DEFAULT_INTERVAL,
        dayOfTheMonth: splitExpression[2] ? splitExpression[2].split(",") : DEFAULT_INTERVAL,
        month: splitExpression[3] ? splitExpression[3].split(",") : DEFAULT_INTERVAL,
        dayOfTheWeek: splitExpression[4] ? splitExpression[4].split(",") : DEFAULT_INTERVAL,
      };
    } else {
      this.expression = {
        minute: DEFAULT_INTERVAL,
        hour: DEFAULT_INTERVAL,
        dayOfTheMonth: DEFAULT_INTERVAL,
        month: DEFAULT_INTERVAL,
        dayOfTheWeek: DEFAULT_INTERVAL,
      };
    }
  }

  /**
   * builds a working cron expression based on the state of the cron object
   * @param {!Object} [options] - customize how to build cron string
   * @param {!Boolean} [options.plain=true] - get cron string as it is, otherwise build short cron string.
   * if false: * 13 * 1-6 0,1,2,3,5,6 ---> * 13 * 1-6 0-3,5-6
   * @param {!Boolean} [options.outputWeekdayNames=false] - changes the numbers to 3 letter weekday names.
   * if true: *\/5 9-17/2 * 1-3 1-5 ---> *\/5 *(10-16)/2 * JAN-MAR MON-FRI
   * @param {!Boolean} [options.outputMonthNames=false] - changes the numbers to 3 letter month names.
   * if true: *\/5 9-17/2 * 1-3 1-5 ---> *\/5 *(10-16)/2 * JAN-MAR MON-FRI
   * @param {!Boolean} [options.outputHashes=false] - changes the * to H.
   * if true: *\/5 9-17/2 * 1-3 1-5 ---> H/5 H(10-16)/2 H 1-3 1-5
   * @returns {string} - working cron expression
   */
  public build(options?: { plain: boolean | undefined } & Omit<Cron.Options, "timezone">): string {
    const {
      plain,
    } = options ?? { plain: true };

    const cronString = [
      this.expression.minute.join(','),
      this.expression.hour.join(','),
      this.expression.dayOfTheMonth.join(','),
      this.expression.month.join(','),
      this.expression.dayOfTheWeek.join(','),
    ].join(' ');

    if (plain) {
      return cronString;
    }

    const cronInstance = new Cron(options);
    cronInstance.fromString(cronString);

    return cronInstance.toString();
  }


  /**
   * adds a value to what exists currently (builds)
   * @param {!String} measureOfTime
   * @param {!String} value
   * @throws {Error} if measureOfTime or value fail validation
   */
  public addValue(measureOfTime: keyof Expression, value: string): void {
    CronValidator.validateValue(measureOfTime, value);

    if (this.expression[measureOfTime].length === 1 && this.expression[measureOfTime][0] === '*') {
      this.expression[measureOfTime] = [value];
    } else {
      if (this.expression[measureOfTime].indexOf(value) < 0) {
        this.expression[measureOfTime].push(value);
      }
    }
  };

  /**
   * removes a single explicit value (subtracts)
   * @param {!String} measureOfTime - as you might guess
   * @param {!String} value - the offensive value
   * @throws {Error} if measureOfTime is bogus.
   */
  public removeValue(measureOfTime: keyof Expression, value: string): void {
    if (!this.expression[measureOfTime]) {
      throw new Error(`Invalid measureOfTime: Valid options are: ${MEASURE_OF_TIME_MAP.join(', ')}`);
    }

    if (this.expression[measureOfTime].length === 1 && this.expression[measureOfTime][0] === '*') {
      console.log('The value for "' + measureOfTime + '" is already at the default value of "*" - this is a no-op.');

      return;
    }

    this.expression[measureOfTime] = this.expression[measureOfTime].filter((timeValue) => {
      return value !== timeValue;
    });

    if (!this.expression[measureOfTime].length) {
      this.expression[measureOfTime] = DEFAULT_INTERVAL;
    }
  };

  /**
   * returns the current state of a given measureOfTime
   * @param {!String} measureOfTime one of "minute", "hour", etc
   * @returns {!Array.number} comma separated blah blah
   * @throws {Error} if the measureOfTime is not one of the permitted values.
   */
  public get(measureOfTime: keyof Expression, options: { expand: true }): number[]
  /**
   * returns the current state of a given measureOfTime
   * @param {!String} measureOfTime one of "minute", "hour", etc
   * @returns {!String} comma separated blah blah
   * @throws {Error} if the measureOfTime is not one of the permitted values.
   */
  public get(measureOfTime: keyof Expression, options?: { expand: false }): string
  public get(measureOfTime: keyof Expression, options?: { expand: boolean }): string | number[] {
    if (!this.expression[measureOfTime]) {
      throw new Error(`Invalid measureOfTime: Valid options are: ${MEASURE_OF_TIME_MAP.join(', ')}`);
    }

    const {
      expand,
    } = options ?? { expand: false };

    if (!expand) {
      return this.expression[measureOfTime].join(',');
    }

    const expression = this.getAll({ expand });

    return expression[measureOfTime];
  }

  /**
   * Returns a rich object that describes the current state of the cron expression.
   * @returns {!{
   *  minute: Array.number,
   *  hour: Array.number,
   *  dayOfTheMonth: Array.number,
   *  month: Array.number,
   *  dayOfTheWeek: Array.number,
   * }}
   */
  public getAll(options: { expand: true }): ExpandedExpression
  /**
   * Returns a rich object that describes the current state of the cron expression.
   * @returns {!{
   *  minute: Array.string,
   *  hour: Array.string,
   *  dayOfTheMonth: Array.string,
   *  month: Array.string,
   *  dayOfTheWeek: Array.string,
   * }}
   */
  public getAll(options?: { expand: false }): Expression
  public getAll(options?: { expand: boolean }): Expression | ExpandedExpression {
    const {
      expand,
    } = options ?? { expand: false };

    if (!expand) {
      return this.expression;
    }

    const cronString = this.build({ plain: true });
    const cronInstance = new Cron();
    cronInstance.fromString(cronString);
    const cronArray = cronInstance.toArray();

    return {
      minute: cronArray[0],
      hour: cronArray[1],
      dayOfTheMonth: cronArray[2],
      month: cronArray[3],
      dayOfTheWeek: cronArray[4],
    }
  }

  /**
   * sets the state of a given measureOfTime
   * @param {!String} measureOfTime - yup
   * @param {!Array.<String>} value - the 5 tuple array of values to set
   * @returns {!String} the comma separated version of the value that you passed in
   * @throws {Error} if your "value" is not an Array&lt;String&gt;
   * @throws {Error} when any item in your value isn't a legal cron-ish descriptor
   */
  public set(measureOfTime: keyof Expression, value: string[]): string {
    if (!Array.isArray(value)) {
      throw new Error('Invalid value; Value must be in the form of an Array.');
    }

    for (const item of value) {
      CronValidator.validateValue(measureOfTime, item);
    }

    this.expression[measureOfTime] = value;

    return this.expression[measureOfTime].join(',');
  }

  /**
   * sets the state for the entire cron expression
   * @param {!{
   *  minute: Array.string,
   *  hour: Array.string,
   *  dayOfTheMonth: Array.string,
   *  month: Array.string,
   *  dayOfTheWeek: Array.string,
   * }} expToSet - the entirety of the cron expression.
   * @throws {Error} as usual
   */
  public setAll(expToSet: Expression): void {
    CronValidator.validateExpression(expToSet);

    keys(this.expression).forEach((key) => this.expression[key] = expToSet[key])
  }
}
